test_guide.md

```diff
@@ -0,0 +1,139 @@
+
+- It exits by calling `process.exit(code)` where `code != 0`
+- It exits due to an uncaught exception.
+- It never exits. In this case, the test runner will terminate the test because it sets a maximum time limit.
+
+They can be added for multiple reasons:
+
+- When adding new functionality.
+- When fixing regression and bugs.
+- Expanding test coverage.
+
+
+## Test structure
+
+Let's analyze this very basic test from the Node.js test suite:
+
+```javascript
+1  'use strict'; 
+2  const common = require('../common');
+3  const http = require('http');
+4  const assert = require('assert');
+5
+6  const server = http.createServer(common.mustCall((req, res) => {
+7    res.end('ok');
+8  }));
+9  server.listen(common.PORT, () => {
+10   http.get({
+11     port: common.PORT,
+12     headers: {'Test': 'Düsseldorf'}
+13   }, common.mustCall((res) => {
+14     assert.equal(res.statusCode, 200);
+15     server.close();
+16   }));
+17 });
+```
+
+**Lines 1-2**
+
+```
+'use strict';
+const common = require('../common');
+```
+
+These 2 lines are mandatory and should be included on every test.
+The `common` module is a helper module that provides useful tools for the tests.
+If for some reason, no functionality from `common` is used, it should still be included like this:
+
+```
+require('../common');
+````
+
+Why? It checks for leaks of globals.
+
+**Lines 3-4**
+
+```
+const http = require('http');
+const assert = require('assert');
+```
+
+These modules are required for the test to run. Except for special cases, these modules should only include the core modules.
+The `assert` module will be used by most of the tests to check that the assumptions for the test are met.
+
+**Lines 6-17**
+
+It's the body of the test. This test is quite simple, it just tests that an HTTP server accepts `obs-text` characters in the headers of an incoming request. Interesting things to notice:
+
+- The use of `common.PORT` as listening port. Always use `common.PORT` instead of using a specific value, as it allows to run tests in parallel safely: they are not trying to reuse the same port another test is already using.
+- The use of `common.mustCall` to check that some callbacks/listeners are called.
+- The http server is closed once all the checks have run. This way, the test can exit gracefully. Remember that for a test to succeed, it must exit with `code == 0`.
+
+## General recommendations
+
+- The use of timers is discouraged, unless we're testing timers. The reasons for this are multiple. Mainly, they are a source of flakiness. For a thorough explanation go [here](https://github.com/nodejs/testing/issues/27).
+
+- Make use of the helpers from the `common` module as much as possible.
+
+  One interesting case is `common.mustCall`. The use `common.mustCall` may avoid the use of extra variables and the corresponding assertions. Let's explain this with a real test from the test suite.
+
+  The following test is `test/parallel/test-http-agent-null.js`
```

Comment:
Maybe don't call out the exact test name, as the name and contents of the file could change.